src/query/storage/promremote/storage.go

```diff
@@ -216,62 +227,94 @@ func (p *promStorage) getTenant(query *storage.WriteQuery) tenant {
 	}
 }
 
-func (p *promStorage) startAsync(ctx context.Context) {
-	p.logger.Info("Start prometheus remote write storage async job",
-		zap.Int("queueSize", p.opts.queueSize),
-		zap.Int("poolSize", p.opts.poolSize))
+func (p *promStorage) flushPendingQueues(minQueueSizeToFlush int, ctx context.Context, wg *sync.WaitGroup) int {
+	numWrites := 0
+	for t, queue := range p.pendingQuery {
+		if queue.Len() == 0 {
+			continue
+		}
+		if queue.Len() < minQueueSizeToFlush {
+			p.logger.Warn("don't do tick flush for small batch",
+				zap.String("tenant", t.name),
+				zap.Int("size", queue.Len()),
+				zap.Int("queue size", p.opts.queueSize))
+			continue
+		}
+		numWrites += queue.Len()
+		wg.Add(1)
+		// Copy the loop variable
+		q := queue
+		p.workerPool.Go(func() {
+			q.Flush(ctx, p)
+			wg.Done()
+		})
+	}
+	return numWrites
+}
+
+func (p *promStorage) writeLoop() {
+	ctxForWrites, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// This loop ensures that all pending writes are flushed before exiting and after `ctx` is done.
+	var wg sync.WaitGroup
 	p.workerPool.Init()
 	ticker := time.NewTicker(*p.opts.tickDuration)
-	go func() {
-		for {
-			select {
-			case query := <-p.queryQueue:
-				if query == nil {
-					time.Sleep(100 * time.Millisecond)
-					continue
-				}
-				t := p.getTenant(query)
-				if _, ok := p.pendingQuery[t]; !ok {
-					p.noTenantFound.Inc(1)
-					p.droppedWrites.Inc(1)
-					p.logger.Error("no pre-defined tenant found, dropping it",
-						zap.String("tenant", t.name), zap.Any("attributes", t.attr),
-						zap.String("defaultTenant", p.opts.tenantDefault),
-						zap.String("timeseries", query.String()))
-					continue
-				}
-				if dataBatch := p.pendingQuery[t].Add(query); dataBatch != nil {
-					p.batchWrite.Inc(int64(len(dataBatch)))
-					p.workerPool.Go(func() {
-						if err := p.writeBatch(ctx, t, dataBatch); err != nil {
-							p.logger.Error("error writing async batch",
-								zap.String("tenant", t.name),
-								zap.Error(err))
-						}
-					})
-				}
-			case <-ticker.C:
-				for t, queue := range p.pendingQuery {
-					if queue.Len() <= p.opts.queueSize/10 {
-						if queue.Len() != 0 {
-							p.logger.Warn("don't do tick flush for small batch",
-								zap.String("tenant", t.name),
-								zap.Int("size", queue.Len()),
-								zap.Int("queue size", p.opts.queueSize))
-						}
-						continue
+	stop := false
+	for !stop {
+		select {
+		case query := <-p.queryQueue:
+			if query == nil {
+				p.logger.Info("Got the poison pill. Exiting the write loop.")
+				// The channel is closed. We should exit.
+				stop = true
```

Comment:
actually we might want to not stop this loop if we got nil, somehow i recall we did got nil :3d-thinking:

Reply:
It shouldn't get nil unless it's closed. I'll test it out.

Reply:
This place is the only one where a sender sends a value to the channel.
<img width="766" alt="image" src="https://github.com/databricks/m3/assets/129140017/19918f9b-d351-4a25-ab12-2841ba55de03">